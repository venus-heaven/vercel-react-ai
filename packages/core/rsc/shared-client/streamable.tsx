import { startTransition, useLayoutEffect, useState } from 'react';
import { STREAMABLE_VALUE_TYPE } from '../constants';
import type { StreamableValue } from '../types';

function hasReadableValueSignature(value: unknown): value is StreamableValue {
  return !!(
    value &&
    typeof value === 'object' &&
    'type' in value &&
    value.type === STREAMABLE_VALUE_TYPE
  );
}

function assertStreamableValue(
  value: unknown,
): asserts value is StreamableValue {
  if (!hasReadableValueSignature(value)) {
    throw new Error(
      'Invalid value: this hook only accepts values created via `createStreamableValue`.',
    );
  }
}

function isStreamableValue(value: unknown): value is StreamableValue {
  const hasSignature = hasReadableValueSignature(value);

  if (!hasSignature && typeof value !== 'undefined') {
    throw new Error(
      'Invalid value: this hook only accepts values created via `createStreamableValue`.',
    );
  }

  return hasSignature;
}

/**
 * `readStreamableValue` takes a streamable value created via the `createStreamableValue().value` API,
 * and returns an async iterator.
 *
 * ```js
 * // Inside your AI action:
 *
 * async function action() {
 *   'use server'
 *   const streamable = createStreamableValue();
 *
 *   streamable.update(1);
 *   streamable.update(2);
 *   streamable.done(3);
 *   // ...
 *   return streamable.value;
 * }
 * ```
 *
 * And to read the value:
 *
 * ```js
 * const streamableValue = await action()
 * for await (const v of readStreamableValue(streamableValue)) {
 *   console.log(v)
 * }
 * ```
 *
 * This logs out 1, 2, 3 on console.
 */
export function readStreamableValue<T = unknown>(
  streamableValue: StreamableValue<T>,
): AsyncIterable<T | undefined> {
  assertStreamableValue(streamableValue);

  return {
    [Symbol.asyncIterator]() {
      let row: StreamableValue<T> | Promise<StreamableValue<T>> =
        streamableValue;
      let curr = row.curr;
      let done = false;
      let initial = true;

      return {
        async next() {
          if (done) return { value: curr, done: true };

          row = await row;

          if (typeof row.error !== 'undefined') {
            throw row.error;
          }
          if ('curr' in row || row.diff) {
            if (row.diff) {
              switch (row.diff[0]) {
                case 0:
                  if (typeof curr !== 'string') {
                    throw new Error(
                      'Invalid patch: can only append to string types. This is a bug in the AI SDK.',
                    );
                  } else {
                    (curr as string) = curr + row.diff[1];
                  }
                  break;
              }
            } else {
              curr = row.curr;
            }

            // The last emitted { done: true } won't be used as the value
            // by the for await...of syntax.
            if (!row.next) {
              done = true;
              return {
                value: curr,
                done: false,
              };
            }
          }

          if (!row.next) {
            return {
              value: curr,
              done: true,
            };
          }

          row = row.next;
          if (initial) {
            initial = false;
            if (typeof curr === 'undefined') {
              // This is the initial chunk and there isn't an initial value yet.
              // Let's skip this one.
              return this.next();
            }
          }

          return {
            value: curr,
            done: false,
          };
        },
      };
    },
  };
}

/**
 * `useStreamableValue` is a React hook that takes a streamable value created via the `createStreamableValue().value` API,
 * and returns the current value, error, and pending state.
 *
 * This is useful for consuming streamable values received from a component's props. For example:
 *
 * ```js
 * function MyComponent({ streamableValue }) {
 *   const [data, error, pending] = useStreamableValue(streamableValue);
 *
 *   if (pending) return <div>Loading...</div>;
 *   if (error) return <div>Error: {error.message}</div>;
 *
 *   return <div>Data: {data}</div>;
 * }
 * ```
 */
export function useStreamableValue<T = unknown, Error = unknown>(
  streamableValue?: StreamableValue<T>,
): [data: T | undefined, error: Error | undefined, pending: boolean] {
  const [curr, setCurr] = useState<T | undefined>(
    isStreamableValue(streamableValue) ? streamableValue.curr : undefined,
  );
  const [error, setError] = useState<Error | undefined>(
    isStreamableValue(streamableValue) ? streamableValue.error : undefined,
  );
  const [pending, setPending] = useState<boolean>(
    isStreamableValue(streamableValue) ? !!streamableValue.next : false,
  );

  useLayoutEffect(() => {
    if (!isStreamableValue(streamableValue)) return;

    let cancelled = false;

    const iterator = readStreamableValue(streamableValue);
    if (streamableValue.next) {
      startTransition(() => {
        if (cancelled) return;
        setPending(true);
      });
    }

    (async () => {
      try {
        for await (const value of iterator) {
          if (cancelled) return;
          startTransition(() => {
            if (cancelled) return;
            setCurr(value);
          });
        }
      } catch (e) {
        if (cancelled) return;
        startTransition(() => {
          if (cancelled) return;
          setError(e as Error);
        });
      } finally {
        if (cancelled) return;
        startTransition(() => {
          if (cancelled) return;
          setPending(false);
        });
      }
    })();

    return () => {
      cancelled = true;
    };
  }, [streamableValue]);

  return [curr, error, pending];
}
