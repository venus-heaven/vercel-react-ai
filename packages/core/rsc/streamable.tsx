import type { ReactNode } from 'react';
import type OpenAI from 'openai';
import { z } from 'zod';
import zodToJsonSchema from 'zod-to-json-schema';

// TODO: This needs to be externalized.
import { OpenAIStream } from '../streams';

import {
  STREAMABLE_VALUE_TYPE,
  DEV_DEFAULT_STREAMABLE_WARNING_TIME,
} from './constants';
import {
  createResolvablePromise,
  createSuspensedChunk,
  consumeStream,
} from './utils';
import type { StreamableValue } from './types';

/**
 * Create a piece of changable UI that can be streamed to the client.
 * On the client side, it can be rendered as a normal React node.
 */
export function createStreamableUI(initialValue?: React.ReactNode) {
  let currentValue = initialValue;
  let closed = false;
  let { row, resolve, reject } = createSuspensedChunk(initialValue);

  function assertStream(method: string) {
    if (closed) {
      throw new Error(method + ': UI stream is already closed.');
    }
  }

  let warningTimeout: NodeJS.Timeout | undefined;
  function warnUnclosedStream() {
    if (process.env.NODE_ENV === 'development') {
      if (warningTimeout) {
        clearTimeout(warningTimeout);
      }
      warningTimeout = setTimeout(() => {
        console.warn(
          'The streamable UI has been slow to update. This may be a bug or a performance issue or you forgot to call `.done()`.',
        );
      }, DEV_DEFAULT_STREAMABLE_WARNING_TIME);
    }
  }
  warnUnclosedStream();

  return {
    value: row,
    update(value: React.ReactNode) {
      assertStream('.update()');

      const resolvable = createResolvablePromise();
      currentValue = value;

      resolve({ value: currentValue, done: false, next: resolvable.promise });
      resolve = resolvable.resolve;
      reject = resolvable.reject;

      warnUnclosedStream();
    },
    append(value: React.ReactNode) {
      assertStream('.append()');

      const resolvable = createResolvablePromise();
      if (typeof currentValue === 'string' && typeof value === 'string') {
        currentValue += value;
      } else {
        currentValue = (
          <>
            {currentValue}
            {value}
          </>
        );
      }

      resolve({ value: currentValue, done: false, next: resolvable.promise });
      resolve = resolvable.resolve;
      reject = resolvable.reject;

      warnUnclosedStream();
    },
    error(error: any) {
      assertStream('.error()');

      if (warningTimeout) {
        clearTimeout(warningTimeout);
      }
      closed = true;
      reject(error);
    },
    done(...args: any) {
      assertStream('.done()');

      if (warningTimeout) {
        clearTimeout(warningTimeout);
      }
      closed = true;
      if (args.length) {
        resolve({ value: args[0], done: true });
        return;
      }
      resolve({ value: currentValue, done: true });
    },
  };
}

/**
 * Create a wrapped, changable value that can be streamed to the client.
 * On the client side, the value can be accessed via the readStreamableValue() API.
 */
export function createStreamableValue<T = any, E = any>(initialValue?: T) {
  let closed = false;
  let { promise, resolve } = createResolvablePromise<StreamableValue<T, E>>();

  function assertStream(method: string) {
    if (closed) {
      throw new Error(method + ': Value stream is already closed.');
    }
  }

  let warningTimeout: NodeJS.Timeout | undefined;
  function warnUnclosedStream() {
    if (process.env.NODE_ENV === 'development') {
      if (warningTimeout) {
        clearTimeout(warningTimeout);
      }
      warningTimeout = setTimeout(() => {
        console.warn(
          'The streamable UI has been slow to update. This may be a bug or a performance issue or you forgot to call `.done()`.',
        );
      }, DEV_DEFAULT_STREAMABLE_WARNING_TIME);
    }
  }
  warnUnclosedStream();

  function createWrapped(
    val: T | undefined,
    initial?: boolean,
  ): StreamableValue<T, E> {
    if (initial) {
      return {
        type: STREAMABLE_VALUE_TYPE,
        curr: val,
        next: promise,
      };
    }

    return {
      curr: val,
      next: promise,
    };
  }

  return {
    value: createWrapped(initialValue, true),
    update(value: T) {
      assertStream('.update()');

      const resolvePrevious = resolve;
      const resolvable = createResolvablePromise();
      promise = resolvable.promise;
      resolve = resolvable.resolve;

      resolvePrevious(createWrapped(value));

      warnUnclosedStream();
    },
    error(error: any) {
      assertStream('.error()');

      if (warningTimeout) {
        clearTimeout(warningTimeout);
      }
      closed = true;
      resolve({ error });
    },
    done(...args: any) {
      assertStream('.done()');

      if (warningTimeout) {
        clearTimeout(warningTimeout);
      }
      closed = true;

      if (args.length) {
        resolve({ curr: args[0] });
        return;
      }

      resolve({});
    },
  };
}

type Streamable = ReactNode | Promise<ReactNode>;
type Renderer<T> = (
  props: T,
) =>
  | Streamable
  | Generator<Streamable, Streamable, void>
  | AsyncGenerator<Streamable, Streamable, void>;

/**
 * `render` is a helper function to create a streamable UI from some LLMs.
 * Currently, it only supports OpenAI's GPT models with Function Calling and Assistants Tools.
 */
export function render<
  TS extends {
    [name: string]: z.Schema;
  } = {},
  FS extends {
    [name: string]: z.Schema;
  } = {},
>(options: {
  /**
   * The model name to use. Must be OpenAI SDK compatible. Tools and Functions are only supported
   * GPT models (3.5/4), OpenAI Assistants, Mistral small and large, and Fireworks firefunction-v1.
   *
   * @example "gpt-3.5-turbo"
   */
  model: string;
  /**
   * The provider instance to use. Currently the only provider available is OpenAI.
   * This needs to match the model name.
   */
  provider: OpenAI;
  messages: Parameters<
    typeof OpenAI.prototype.chat.completions.create
  >[0]['messages'];
  text?: Renderer<{ content: string; done: boolean }>;
  tools?: {
    [name in keyof TS]: {
      description?: string;
      parameters: TS[name];
      render: Renderer<z.infer<TS[name]>>;
    };
  };
  functions?: {
    [name in keyof FS]: {
      description?: string;
      parameters: FS[name];
      render: Renderer<z.infer<FS[name]>>;
    };
  };
  initial?: ReactNode;
  temperature?: number;
}): ReactNode {
  const ui = createStreamableUI(options.initial);

  // The default text renderer just returns the content as string.
  const text = options.text
    ? options.text
    : ({ content }: { content: string }) => content;

  const functions = options.functions
    ? Object.entries(options.functions).map(
        ([name, { description, parameters }]) => {
          return {
            name,
            description,
            parameters: zodToJsonSchema(parameters) as Record<string, unknown>,
          };
        },
      )
    : undefined;

  const tools = options.tools
    ? Object.entries(options.tools).map(
        ([name, { description, parameters }]) => {
          return {
            type: 'function' as const,
            function: {
              name,
              description,
              parameters: zodToJsonSchema(parameters) as Record<
                string,
                unknown
              >,
            },
          };
        },
      )
    : undefined;

  if (functions && tools) {
    throw new Error(
      "You can't have both functions and tools defined. Please choose one or the other.",
    );
  }

  let finished: Promise<void> | undefined;

  async function handleRender(
    args: any,
    renderer: undefined | Renderer<any>,
    res: ReturnType<typeof createStreamableUI>,
  ) {
    if (!renderer) return;

    const resolvable = createResolvablePromise<void>();

    if (finished) {
      finished = finished.then(() => resolvable.promise);
    } else {
      finished = resolvable.promise;
    }

    const value = renderer(args);
    if (
      value instanceof Promise ||
      (value &&
        typeof value === 'object' &&
        'then' in value &&
        typeof value.then === 'function')
    ) {
      const node = await (value as Promise<React.ReactNode>);
      res.update(node);
      resolvable.resolve(void 0);
    } else if (
      value &&
      typeof value === 'object' &&
      Symbol.asyncIterator in value
    ) {
      const it = value as AsyncGenerator<
        React.ReactNode,
        React.ReactNode,
        void
      >;
      while (true) {
        const { done, value } = await it.next();
        res.update(value);
        if (done) break;
      }
      resolvable.resolve(void 0);
    } else if (value && typeof value === 'object' && Symbol.iterator in value) {
      const it = value as Generator<React.ReactNode, React.ReactNode, void>;
      while (true) {
        const { done, value } = it.next();
        res.update(value);
        if (done) break;
      }
      resolvable.resolve(void 0);
    } else {
      res.update(value);
      resolvable.resolve(void 0);
    }
  }

  (async () => {
    let hasFunction = false;
    let content = '';

    consumeStream(
      OpenAIStream(
        (await options.provider.chat.completions.create({
          model: options.model,
          messages: options.messages,
          temperature: options.temperature,
          stream: true,
          ...(functions
            ? {
                functions,
              }
            : {}),
          ...(tools
            ? {
                tools,
              }
            : {}),
        })) as any,
        {
          ...(functions
            ? {
                async experimental_onFunctionCall(functionCallPayload) {
                  hasFunction = true;
                  handleRender(
                    functionCallPayload.arguments,
                    options.functions?.[functionCallPayload.name as any]
                      ?.render,
                    ui,
                  );
                },
              }
            : {}),
          ...(tools
            ? {
                async experimental_onToolCall(toolCallPayload: any) {
                  hasFunction = true;

                  // TODO: We might need Promise.all here?
                  for (const tool of toolCallPayload.tools) {
                    handleRender(
                      tool.func.arguments,
                      options.tools?.[tool.func.name as any]?.render,
                      ui,
                    );
                  }
                },
              }
            : {}),
          onText(chunk) {
            content += chunk;
            handleRender({ content, done: false }, text, ui);
          },
          async onFinal() {
            if (hasFunction) {
              await finished;
              ui.done();
              return;
            }

            handleRender({ content, done: true }, text, ui);
            await finished;
            ui.done();
          },
        },
      ),
    );
  })();

  return ui.value;
}
