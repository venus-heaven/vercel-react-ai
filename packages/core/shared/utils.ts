import {
  StreamPartType,
  StreamStringPrefixes,
  parseStreamPart,
} from './stream-parts';

export * from './generate-id';

// TODO remove (breaking change)
export { generateId as nanoid } from './generate-id';

// Export stream data utilities for custom stream implementations,
// both on the client and server side.
export type { StreamPart } from './stream-parts';
export { formatStreamPart, parseStreamPart } from './stream-parts';
export { readDataStream } from './read-data-stream';

// simple decoder signatures:
function createChunkDecoder(): (chunk: Uint8Array | undefined) => string;
function createChunkDecoder(
  complex: false,
): (chunk: Uint8Array | undefined) => string;
// complex decoder signature:
function createChunkDecoder(
  complex: true,
): (chunk: Uint8Array | undefined) => StreamPartType[];
// combined signature for when the client calls this function with a boolean:
function createChunkDecoder(
  complex?: boolean,
): (chunk: Uint8Array | undefined) => StreamPartType[] | string;
function createChunkDecoder(complex?: boolean) {
  const decoder = new TextDecoder();

  if (!complex) {
    return function (chunk: Uint8Array | undefined): string {
      if (!chunk) return '';
      return decoder.decode(chunk, { stream: true });
    };
  }

  return function (chunk: Uint8Array | undefined) {
    const decoded = decoder
      .decode(chunk, { stream: true })
      .split('\n')
      .filter(line => line !== ''); // splitting leaves an empty string at the end

    return decoded.map(parseStreamPart).filter(Boolean);
  };
}

export { createChunkDecoder };

export const isStreamStringEqualToType = (
  type: keyof typeof StreamStringPrefixes,
  value: string,
): value is StreamString =>
  value.startsWith(`${StreamStringPrefixes[type]}:`) && value.endsWith('\n');

export type StreamString =
  `${(typeof StreamStringPrefixes)[keyof typeof StreamStringPrefixes]}:${string}\n`;
