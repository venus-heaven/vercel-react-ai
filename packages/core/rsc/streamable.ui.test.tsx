import {
  openaiChatCompletionChunks,
  openaiFunctionCallChunks,
} from '../tests/snapshots/openai-chat';
import { DEFAULT_TEST_URL, createMockServer } from '../tests/utils/mock-server';
import { createStreamableUI, render } from './streamable';
import { z } from 'zod';

const FUNCTION_CALL_TEST_URL = DEFAULT_TEST_URL + 'mock-func-call';

// This is a workaround to render the Flight response in a test environment.
async function flightRender(node: React.ReactNode, byChunk?: boolean) {
  const ReactDOM = require('react-dom');
  ReactDOM.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactDOMCurrentDispatcher =
    { current: {} };

  const React = require('react');
  React.__SECRET_SERVER_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED = {
    ReactSharedServerInternals: {},
    ReactCurrentCache: {
      current: null,
    },
  };

  const {
    renderToReadableStream,
  } = require('react-server-dom-webpack/server.edge');

  const stream = renderToReadableStream(node);
  const reader = stream.getReader();

  const chunks = [];
  let result = '';
  while (true) {
    const { done, value } = await reader.read();
    if (done) {
      break;
    }

    const decoded = new TextDecoder().decode(value);
    if (byChunk) {
      chunks.push(decoded);
    } else {
      result += decoded;
    }
  }

  return byChunk ? chunks : result;
}

const server = createMockServer([
  {
    url: DEFAULT_TEST_URL,
    chunks: openaiChatCompletionChunks,
    formatChunk: chunk => `data: ${JSON.stringify(chunk)}\n\n`,
    suffix: 'data: [DONE]',
  },
  {
    url: FUNCTION_CALL_TEST_URL,
    chunks: openaiFunctionCallChunks,
    formatChunk: chunk => `data: ${JSON.stringify(chunk)}\n\n`,
    suffix: 'data: [DONE]',
  },
]);

beforeAll(() => {
  server.listen();
});

afterEach(() => {
  server.resetHandlers();
});

afterAll(() => {
  server.close();
});

function nextTick() {
  return Promise.resolve();
}

async function recursiveResolve(val: any): Promise<any> {
  if (val && typeof val === 'object' && typeof val.then === 'function') {
    return await recursiveResolve(await val);
  }

  if (Array.isArray(val)) {
    return await Promise.all(val.map(recursiveResolve));
  }

  if (val && typeof val === 'object') {
    const result: any = {};
    for (const key in val) {
      result[key] = await recursiveResolve(val[key]);
    }
    return result;
  }

  return val;
}

async function simulateFlightServerRender(node: React.ReactElement) {
  async function traverse(node: any): Promise<any> {
    if (!node) return {};

    // Let's only do one level of promise resolution here. As it's only for testing purposes.
    const props = await recursiveResolve({ ...node.props } || {});

    const { type } = node;
    const { children, ...otherProps } = props;
    const typeName = typeof type === 'function' ? type.name : String(type);

    return {
      type: typeName,
      props: otherProps,
      children:
        typeof children === 'string'
          ? children
          : Array.isArray(children)
          ? children.map(traverse)
          : await traverse(children),
    };
  }

  return traverse(node);
}

function getFinalValueFromResolved(node: any) {
  if (!node) return node;
  if (node.type === 'Symbol(react.suspense)') {
    return getFinalValueFromResolved(node.children);
  } else if (node.type === '') {
    let wrapper;
    let value = node.props.value;
    let next = node.props.n;
    let current = node.props.c;

    while (next) {
      if (next.append) {
        if (wrapper === undefined) {
          wrapper = current;
        } else if (typeof current === 'string' && typeof wrapper === 'string') {
          wrapper = wrapper + current;
        } else {
          wrapper = (
            <>
              {wrapper}
              {current}
            </>
          );
        }
      }

      value = next.value;
      next = next.next;
      current = value;
    }

    return getFinalValueFromResolved(
      wrapper === undefined ? (
        value
      ) : typeof value === 'string' && typeof wrapper === 'string' ? (
        wrapper + value
      ) : (
        <>
          {wrapper}
          {value}
        </>
      ),
    );
  }
  return node;
}

function createMockUpProvider() {
  return {
    chat: {
      completions: {
        create: async () => {
          return await fetch(FUNCTION_CALL_TEST_URL);
        },
      },
    },
  } as any;
}

describe('rsc - render()', () => {
  it('should emit React Nodes with sync render function', async () => {
    const ui = render({
      model: 'gpt-3.5-turbo',
      messages: [],
      provider: createMockUpProvider(),
      functions: {
        get_current_weather: {
          description: 'Get the current weather',
          parameters: z.object({}),
          render: () => {
            return <div>Weather</div>;
          },
        },
      },
    });

    const rendered = await simulateFlightServerRender(ui as any);
    expect(rendered).toMatchSnapshot();
  });

  it('should emit React Nodes with async render function', async () => {
    const ui = render({
      model: 'gpt-3.5-turbo',
      messages: [],
      provider: createMockUpProvider(),
      functions: {
        get_current_weather: {
          description: 'Get the current weather',
          parameters: z.object({}),
          render: async () => {
            await new Promise(resolve => setTimeout(resolve, 100));
            return <div>Weather</div>;
          },
        },
      },
    });

    const rendered = await simulateFlightServerRender(ui as any);
    expect(rendered).toMatchSnapshot();
  });

  it('should emit React Nodes with generator render function', async () => {
    const ui = render({
      model: 'gpt-3.5-turbo',
      messages: [],
      provider: createMockUpProvider(),
      functions: {
        get_current_weather: {
          description: 'Get the current weather',
          parameters: z.object({}),
          render: async function* () {
            yield <div>Loading...</div>;
            await new Promise(resolve => setTimeout(resolve, 100));
            return <div>Weather</div>;
          },
        },
      },
    });

    const rendered = await simulateFlightServerRender(ui as any);
    expect(rendered).toMatchSnapshot();
  });
});

describe('rsc - createStreamableUI()', () => {
  it('should emit React Nodes that can be updated', async () => {
    const ui = createStreamableUI(<div>1</div>);
    ui.update(<div>2</div>);
    ui.update(<div>3</div>);
    ui.done();

    const final = getFinalValueFromResolved(
      await simulateFlightServerRender(ui.value),
    );
    expect(final).toMatchInlineSnapshot(`
      <div>
        3
      </div>
    `);
  });

  it('should emit React Nodes that can be updated with .done()', async () => {
    const ui = createStreamableUI(<div>1</div>);
    ui.update(<div>2</div>);
    ui.update(<div>3</div>);
    ui.done(<div>4</div>);

    const final = getFinalValueFromResolved(
      await simulateFlightServerRender(ui.value),
    );
    expect(final).toMatchInlineSnapshot(`
      <div>
        4
      </div>
    `);
  });

  it('should support .append()', async () => {
    const ui = createStreamableUI(<div>1</div>);
    ui.update(<div>2</div>);
    ui.append(<div>3</div>);
    ui.append(<div>4</div>);
    ui.done();

    const final = getFinalValueFromResolved(
      await simulateFlightServerRender(ui.value),
    );
    expect(final).toMatchInlineSnapshot(`
      <React.Fragment>
        <React.Fragment>
          <div>
            2
          </div>
          <div>
            3
          </div>
        </React.Fragment>
        <div>
          4
        </div>
      </React.Fragment>
    `);
  });

  it('should support streaming .append() result before .done()', async () => {
    const ui = createStreamableUI(<div>1</div>);
    ui.append(<div>2</div>);
    ui.append(<div>3</div>);

    const currentRsolved = ui.value.props.children.props.n;
    const tryResolve1 = await Promise.race([currentRsolved, nextTick()]);
    expect(tryResolve1).toBeDefined();
    const tryResolve2 = await Promise.race([tryResolve1.next, nextTick()]);
    expect(tryResolve2).toBeDefined();
    expect(getFinalValueFromResolved(tryResolve2.value)).toMatchInlineSnapshot(`
      <div>
        3
      </div>
    `);

    ui.append(<div>4</div>);
    ui.done();

    const final = getFinalValueFromResolved(
      await simulateFlightServerRender(ui.value),
    );
    expect(final).toMatchInlineSnapshot(`
      <React.Fragment>
        <React.Fragment>
          <React.Fragment>
            <div>
              1
            </div>
            <div>
              2
            </div>
          </React.Fragment>
          <div>
            3
          </div>
        </React.Fragment>
        <div>
          4
        </div>
      </React.Fragment>
    `);
  });

  it('should support updating the appended ui', async () => {
    const ui = createStreamableUI(<div>1</div>);
    ui.update(<div>2</div>);
    ui.append(<div>3</div>);
    ui.done(<div>4</div>);

    const final = getFinalValueFromResolved(
      await simulateFlightServerRender(ui.value),
    );
    expect(final).toMatchInlineSnapshot(`
      <React.Fragment>
        <div>
          2
        </div>
        <div>
          4
        </div>
      </React.Fragment>
    `);
  });

  it('should re-use the text node when appending strings', async () => {
    const ui = createStreamableUI('hello');
    ui.append(' world');
    ui.append('!');
    ui.done();

    const final = getFinalValueFromResolved(
      await simulateFlightServerRender(ui.value),
    );
    expect(final).toMatchInlineSnapshot('"hello world!"');
  });

  it('should send minimal incremental diffs when appending strings', async () => {
    const ui = createStreamableUI('hello');
    ui.append(' world');
    ui.append(' and');
    ui.append(' universe');
    ui.done();

    expect(await flightRender(ui.value)).toMatchInlineSnapshot(`
      "1:\\"$Sreact.suspense\\"
      2:D{\\"name\\":\\"\\",\\"env\\":\\"Server\\"}
      0:[\\"$\\",\\"$1\\",null,{\\"fallback\\":\\"hello\\",\\"children\\":\\"$L2\\"}]
      3:D{\\"name\\":\\"\\",\\"env\\":\\"Server\\"}
      2:[\\"hello\\",[\\"$\\",\\"$1\\",null,{\\"fallback\\":\\" world\\",\\"children\\":\\"$L3\\"}]]
      4:D{\\"name\\":\\"\\",\\"env\\":\\"Server\\"}
      3:[\\" world\\",[\\"$\\",\\"$1\\",null,{\\"fallback\\":\\" and\\",\\"children\\":\\"$L4\\"}]]
      5:D{\\"name\\":\\"\\",\\"env\\":\\"Server\\"}
      4:[\\" and\\",[\\"$\\",\\"$1\\",null,{\\"fallback\\":\\" universe\\",\\"children\\":\\"$L5\\"}]]
      5:\\" universe\\"
      "
    `);

    const final = getFinalValueFromResolved(
      await simulateFlightServerRender(ui.value),
    );
    expect(final).toMatchInlineSnapshot('"hello world and universe"');
  });

  it('should error when updating a closed streamable', async () => {
    const ui = createStreamableUI(<div>1</div>);
    ui.done(<div>2</div>);

    expect(() => {
      ui.update(<div>3</div>);
    }).toThrowErrorMatchingInlineSnapshot(
      '".update(): UI stream is already closed."',
    );
  });

  it('should avoid sending data again if the same UI is passed', async () => {
    const node = <div>1</div>;
    const ui = createStreamableUI(node);
    ui.update(node);
    ui.update(node);
    ui.update(node);
    ui.update(node);
    ui.update(node);
    ui.update(node);
    ui.done();

    expect(await flightRender(ui.value)).toMatchInlineSnapshot(`
      "1:\\"$Sreact.suspense\\"
      2:D{\\"name\\":\\"\\",\\"env\\":\\"Server\\"}
      0:[\\"$\\",\\"$1\\",null,{\\"fallback\\":[\\"$\\",\\"div\\",null,{\\"children\\":\\"1\\"}],\\"children\\":\\"$L2\\"}]
      4:{\\"children\\":\\"1\\"}
      3:[\\"$\\",\\"div\\",null,\\"$4\\"]
      2:\\"$3\\"
      "
    `);
  });
});
