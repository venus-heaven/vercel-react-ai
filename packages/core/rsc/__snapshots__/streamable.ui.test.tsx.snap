// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rsc - render() > should emit React Nodes with async render function 1`] = `
{
  "children": {
    "children": {},
    "props": {
      "current": undefined,
      "next": {
        "done": false,
        "next": {
          "done": true,
          "value": <div>
            Weather
          </div>,
        },
        "value": <div>
          Weather
        </div>,
      },
    },
    "type": "Row",
  },
  "props": {
    "fallback": undefined,
  },
  "type": "Symbol(react.suspense)",
}
`;

exports[`rsc - render() > should emit React Nodes with generator render function 1`] = `
{
  "children": {
    "children": {},
    "props": {
      "current": undefined,
      "next": {
        "done": false,
        "next": {
          "done": false,
          "next": {
            "done": true,
            "value": <div>
              Weather
            </div>,
          },
          "value": <div>
            Weather
          </div>,
        },
        "value": <div>
          Loading...
        </div>,
      },
    },
    "type": "Row",
  },
  "props": {
    "fallback": undefined,
  },
  "type": "Symbol(react.suspense)",
}
`;

exports[`rsc - render() > should emit React Nodes with sync render function 1`] = `
{
  "children": {
    "children": {},
    "props": {
      "current": undefined,
      "next": {
        "done": false,
        "next": {
          "done": true,
          "value": <div>
            Weather
          </div>,
        },
        "value": <div>
          Weather
        </div>,
      },
    },
    "type": "Row",
  },
  "props": {
    "fallback": undefined,
  },
  "type": "Symbol(react.suspense)",
}
`;
